File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) -1
        while l <= r:
            m = (l+r)//2
            if nums[m] == target:
                return m
            
            if nums[l] <= nums[m]:
                if nums[l] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            
            else:
                if nums[m] < target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1

        #l, r = 0, len(nums) - 1
        
        
        #while l <= r:
            #mid = (l + r) // 2
            
            #if nums[mid] == target:
                #return mid
            
            # 左半部有序
            
            #if nums[l] <= nums[mid]:
                #if nums[l] <= target < nums[mid]:
                    #r = mid - 1
                #else:# 如果左半部呈現無序狀況，代表目標數在右半邊
                    #l = mid + 1
            # 右半部有序            
            #else:
                #if nums[mid] < target <= nums[r]:
                    #l = mid + 1
                #else:
                    #r = mid - 1
        #return -1


                ##[3,4,5,6,1,2]

                ## target = 1 
                ## l = 0 r= 5 mid = 2
                ## nums[2]= 5 
                ## (l+5)/2= 4 l = 3 = m + 1





        
        
        
        

File: Data Structures & Algorithms/sort-colors/submission-2.java
class Solution {
    public void sortColors(int[] nums) {
        HashMap<Integer,Integer> count = new HashMap<>();
        for (int n : nums)
        {
            count.put(n,count.getOrDefault(n,0)+1);
        }

        int[] numList = new int[]{0,1,2};
        int index = 0;
        for (int i = 0; i < numList.length; i++)
        {
            int color = numList[i];

            while(count.getOrDefault(color,0) > 0)
            {
                nums[index] = color;
                index++;
                count.put(color, count.get(color)-1);
            }
        }
    }
}